### PR TITLE
Align development test helpers with 3.6

### PR DESCRIPTION
**Pull request by davidhorstmann-arm**
Add version guards to allow `development`'s test helpers to work in the 3.6 branch, in preparation for moving these helpers to the framework repo.

## PR checklist

Please remove the segment/s on either side of the | symbol as appropriate, and add any relevant link/s to the end of the line.
If the provided content is part of the present PR remove the # symbol.

- [ ] **changelog** provided | not required because: 
- [ ] **development PR** provided # | not required because: 
- [ ] **framework PR** provided Mbed-TLS/mbedtls-framework# | not required
- [ ] **3.6 PR** provided # | not required because: 
- [ ] **2.28 PR** provided # | not required because: 
- **tests**  provided | not required because:

### CODE REVIEW

**Comment by mpg** [comment]
I'll note that currently in development `VERSION_MAJOR` is still `3` and `VERSION_MINOR` is still `6`, so the current condition does not distinguish between development and actual 3.6 - it will evaluate to true in both cases.

So, I'd say if we want to be able to distinguish now, we need to change something in development: we could either immediately switch to 4.0 (then I'd agree we only need to check `VERSION_MAJOR`), or if we're not comfortable with bumping the version number to that value before it's actually done (note: normally, bumping the version is part of the release process), use some intermediate like 3.99 - and then we need to check `VERSION_MINOR` too as the PR is doing.

Or we could even go a different route and not use `MBEDTLS_VERSION_MAJOR/MINOR` but instead have a dedicated macro that's defined only in development or only in the LTS branch. (`MBEDTLS_VERSION_DEVELOPMENT` vs `MBEDTLS_VERSION_36LTS` for example. Or perhaps `MBEDTLS_BRANCH_xxx` instead of `MBEDTLS_VERSION` actually.)

Wdyt?

**Comment by gilles-peskine-arm** [comment]
> we could either immediately switch to 4.0 (then I'd agree we only need to check VERSION_MAJOR)

We've already made backward-incompatible changes in `development`, so arguably we should already have bumped the major version.

**Comment by davidhorstmann-arm** [comment]
@mpg thanks, I missed that and indeed the CI is failing because `development` is detected as 3.6. For now I've bumped the version to 3.9.9 in this PR for testing purposes.

As for the final state of this PR, I tend to think keeping the version as 3.9.9 is the best approach because it keeps our release process as it is and allows us to mark the release with an execution of `bump_version.sh`. However, I see Gilles' point that since we are currently backwards-incompatible with 3.6 it makes no sense for us to have a 3.x version number.

As I've mentioned before, I'm somewhat sceptical about the need for us to move test helpers to the framework at all, since at the moment version control effectively does this per-branch functionality-gating for us, but that is largely a separate discussion which I will raise elsewhere.

**Comment by mpg** [comment]
> We've already made backward-incompatible changes in development, so arguably we should already have bumped the major version.

I'm not convinced the rules of semantic versioning are supposed to apply to unreleased branches. On the other hand, I agree that if we change the version number, it feels weird not to change the major component while there are already incompatible changes.

I'm uncomfortable making an exception to the release process by bumping the version to its final value in advance - processes work best when there are no exceptions.

I'm really tempted to not touch the version at all, and rely on a new dedicated macro instead. I don't think this new macro needs to be public, it could live in `library/common.h`. Wdyt?

**Comment by ronald-cron-arm** [comment]
> I'm really tempted to not touch the version at all, and rely on a new dedicated macro instead. I don't think this new macro needs to be public, it could live in `library/common.h`. Wdyt?

I am not a fan of the 3.99 trick thus rather in favor of a dedicated macro solution.

File: tests/src/certs.c [hunk 1]
```diff
@@ -301,13 +301,15 @@ const size_t mbedtls_test_cli_crt_ec_len =
  * Dispatch between SHA-1 and SHA-256
  */
 
-#if defined(PSA_WANT_ALG_SHA_256)
+#if ((MBEDTLS_VERSION_MAJOR == 3) && (MBEDTLS_VERSION_MINOR == 6) \
```
Comment (ronald-cron-arm) [inline tests/src/certs.c:304]: Another possibility would be to restore `MBEDTLS_MD_CAN_SHA256` here and in a development test header to have something like:
```C
#if defined(PSA_WANT_ALG_SHA256)
#define MBEDTLS_MD_CAN_SHA256
#endif
```
Wdyt?

Comment (ronald-cron-arm) [reply tests/src/certs.c:304]: Other possibility, we keep this file version dependent?

File: tests/src/drivers/hash.c [hunk 2]
```diff
@@ -13,8 +13,12 @@
 #include "test/drivers/hash.h"
 
 #if defined(MBEDTLS_TEST_LIBTESTDRIVER1)
+#if (MBEDTLS_VERSION_MAJOR == 3) && (MBEDTLS_VERSION_MINOR == 6)
```
Comment (ronald-cron-arm) [inline tests/src/drivers/hash.c:16]: We will revisit that when we migrate the libtestdriver1 support to TF-PSA-Crypto and CMake. We may be able to get rid of the `#if (MBEDTLS_VERSION_MAJOR == 3) && (MBEDTLS_VERSION_MINOR == 6)` or what will replace it in this PR. Added a note `to https://github.com/Mbed-TLS/TF-PSA-Crypto/issues/61`.

**Comment by gilles-peskine-arm** [comment]
> I'm uncomfortable making an exception to the release process by bumping the version to its final value in advance - processes work best when there are no exceptions.

1. I think it would actually make more sense to change our process to bump the versions (both API and ABI) the first time we make a change that departs from the previous release
2. We don't really have a process for major releases.

So I still push for changing the major version to 4 yesterday.

**Comment by mpg** [comment]
If we want to discuss changing our process about bumping version numbers, then that discussion needs a larger audience than just the people who follow this PR.

In the meantime, using a dedicated macro would unblock this PR by disentangling it from a larger process discussion.

File: tests/src/certs.c (hunk 1 above)
Comment (mpg) [reply tests/src/certs.c:304]: I said in the meeting that I though it would be weird to keep this file version-dependent while it's essentially a C version of files that are in the framework. Turns out I was confused: I was thinking about `test_certs.h` not `certs.c`. So, I'm less opposed to keeping this file version-dependent than I was in the meeting.

Still, what I wanted to suggest is that we could have helper macros, but I'd rather give them different names than the 3.6 macros, just for clarity - to make sure we don't accidentally keep using the old macros when we're not supposed to. The macros could be local to this file and be like `CERTS_HAVE_SHA256` etc.

Just a suggestion, now that I've realised the difference between `test_certs.h` and `certs.c` maybe keeping `certs.c` version-dependent is a good option after all.

Comment (gilles-peskine-arm) [reply tests/src/certs.c:304]: We're likely to need this in several places in the test data. Why would we have different macros in each file?

Regarding this particular file though, do we actually need to prevent unsupported certificates from being defined? This is test data so code size is not really relevant. Why is it bad if we define constants that just don't happen to be used in a particular build?

Comment (mpg) [reply tests/src/certs.c:304]: > Regarding this particular file though, do we actually need to prevent unsupported certificates from being defined?

Comment (mpg) [reply tests/src/certs.c:304]: > We're likely to need this in several places in the test data. Why would we have different macros in each file?

Indeed. We could have one version-dependent file with a bunch of things like, in development:
```
#if defined(PSA_WANT_ALG_SHA_256)
#define MBEDTLS_TEST_MD_CAN_SHA256
#define MBEDTLS_TEST_SHA256_C
#endif
```
and in 3.6:
```
#if defined(MBEDTLS_MD_CAN_SHA256)
#define MBEDTLS_TEST_MD_CAN_SHA256
#endif
#if defined(MBEDTLS_SHA256_C)
#define MBEDTLS_TEST_SHA256_C
#endif
```
and then everywhere else we just use the `MBEDTLS_TEST_xxx` macros. (Those need to be named after the 3.6 macros, not `PSA_WANT`, otherwise we'd loose information that's useful in 3.6.)

Then we don't even need to explicitly check for the version number anywhere, because we just use those macros and they are defined in a file that's version-dependent so we automagically get the correct definitions (or lack of definition) for the current version.

Wdyt?

**Comment by davidhorstmann-arm** [comment]
Re versioning, I can see the argument for bumping the version before a release.

On these specific cases though, having done a bit more thinking I now think that it would be best to try and find a more localistic solution to each of these 3.6 - 4.x conflicts that will allow us to avoid:
```c
#if IS_VERSION_3_6
/* Do things the ye olde way */
#else
/* Do things the 4.x way */
#endif
```
In many of these cases we can just accept either the old or new config option, without worrying about which branch it comes from. For the libtestdriver differences, I think there may be an ecumenical way to address this with some careful macro work - I'll have a go and report back as to how I get on.

File: tests/src/certs.c (hunk 1 above)
Comment (davidhorstmann-arm) [reply tests/src/certs.c:304]: That seems like a good approach to me, since it solves the essential problem of having version-dependent code in a version-independent repo. This could also work well for the libtestdriver paths:
```c
#define MBEDTLS_TEST_DRIVER_HEADER_DIR "libtestdriver1/tf-psa-crypto/core"
```
and so on.

Comment (tom-cosgrove-arm) [reply tests/src/certs.c:304]: The point of the framework is to share common code. There doesn’t seem to be any commonality in the lines in the PR here - it’s an interleaving of 3.6 and 4.x code, so is really calling to be version-dependent.

`MBEDTLS_TEST_MD_CAN_SHA256 ` &cc seems to be adding an additional layer of indirection: how are we going to test this? Test code needs to be obviously correct, otherwise we need to start testing the test code, then testing the code that tests the test code, then...

(For an example of testing the tests, it's so hard to fathom if we actually do run all the tests that we've written a script `analyze_outcomes.py` to do this for us. That in itself has had bugs, so...)

File: tests/include/test/ssl_helpers.h [hunk 3]
```diff
@@ -38,21 +38,26 @@
...
 #endif
-#if defined(PSA_WANT_ALG_SHA_256)
+#if defined(MBEDTLS_MD_CAN_SHA256) \
+    || defined(PSA_WANT_ALG_SHA256)
 #define MBEDTLS_TEST_HAS_TLS1_3_AES_128_GCM_SHA256
 #endif
-#endif /* PSA_WANT_ALG_GCM */
-#if defined(PSA_WANT_ALG_CCM) && defined(PSA_WANT_ALG_SHA_256)
+#endif /* MBEDTLS_SSL_HAVE_GCM || PSA_WANT_ALG_GCM */
+#if defined(PSA_WANT_ALG_CCM) \
```
[the remark this answers is not in the document]
Comment (davidhorstmann-arm) [reply tests/include/test/ssl_helpers.h:51]: Oop, no you are correct, I had missed that one and some others, thanks!

**Comment by mpg** [comment]
@davidhorstmann-arm We've been discussing this with Ronald, and came to the conclusion that we'd like to reduce the (initial) scope here to only include what's needed for the repo split, it only files used by the crypto part. Practically, that would exclude `certs.c` and `ssl_helpers.*`.

The reasoning is that there are things that depend on the crypto part being done, but nothing that depends on the X.509/TLS part, which also happens to be where most of the difficult problems live. So, we'd like to avoid delaying the urgent-and-relatively-easy part due to the not-urgent-and-harder part.

Could you update this and #9547 with that new reduced scope in mind? Thanks!

**Comment by davidhorstmann-arm** [comment]
@mpg I have de-scoped both PRs and I expect CI to pass (it was passing with the larger scope). This and #9547 should be ready for review.

**Review by eleuzi01 — APPROVED** [review]
LGTM, thanks!

**Review by mpg — COMMENTED** [review]
The changes look correct to me.

However I found  the preprocessor conditions in `psa_crypto_helpers.h` pretty hard to follow (even compared to our average, which is already too high). I'm not sure what can be done to make the situation better, though, but let me float an idea and let's discuss it.

From my perspective the file has two parts:
1. The various `*PSA_INIT`/`*PSA_DONE` macros and their various definitions. This is where most of the complexity lies, as in (a) 3.6 is very complex and (b) 3.6 and 4.0/1.0 differ substantially.
2. Everything else. There the initial complexity is lower and the differences simpler, in that we have the same number of cases on both sides, just the conditions are expressed differently (CLIENT vs C, and ALT macros that exist only in 3.6).

I'm very tempted to suggest we split the file in two, and only try to share the second part between 3.6 and 1.0.

The first part (`*PSA_INIT/DONE`) could be kept as-is for 3.6 (with all its current complexity, but without the added complexity of unifying with 4.0/1.0), and have a much simpler version shared between 4.0 and 1.0 only - because there we should only need `PSA_INIT/DONE`, not the various `xxx_PSA_INIT/DONE` variants, and there will be fewer cases: PSA is always on (the only question being do we have `C` or only `CLIENT`). Sharing the file between 3.6 and 4.0/1.0 would force use to carry all the complexity of 3.6 in 4.0/1.0 which I don't think is a desirable outcome.

Wdyt?

**Comment by mpg** [comment]
PS: can you update the description to reflect the new reduced scope?

**Comment by davidhorstmann-arm** [comment]
> Sharing the file between 3.6 and 4.0/1.0 would force use to carry all the complexity of 3.6 in 4.0/1.0 which I don't think is a desirable outcome.
>
> Wdyt?

Hmm. I'm not too keen on splitting the header file to straddle 3.6 and 4.0 - that seems like we might be just adding the extra complexity by the backdoor. Couldn't we just wait until `USE_PSA` is removed and then put all of the `*PSA_INIT()` and `*PSA_DONE()` macros behind a big `#if MBEDTLS_VERSION_MAJOR < 4` guard?

**Comment by gilles-peskine-arm** [comment]
> I'm not too keen on splitting the header file to straddle 3.6 and 4.0 - that seems like we might be just adding the extra complexity by the backdoor.

Historical note (I haven't looked at any details): there was never any particular reason for grouping the content of `psa_crypto_helpers.h`, it was just stuff that was somehow related to PSA. We originally had everything in `helpers.function` and gradually moved it out. Splitting cleanup-related declarations, the decision of calling init, support for exotic options and whatever else would be rather cleaning up a bit.

> wait until USE_PSA is removed

I'm not sure it will help that much here, but [it's coming very soon](https://github.com/Mbed-TLS/mbedtls/pull/9614). (And you're welcome to review it if you want to hurry it along...)

**Comment by mpg** [comment]
> Couldn't we just wait until `USE_PSA` is removed and then put all of the `*PSA_INIT()` and `*PSA_DONE()` macros behind a big `#if MBEDTLS_VERSION_MAJOR < 4` guard?

It's not just about `USE_PSA`, see the various other macros, it's also whether MD, CTR-DRBG, `block_cipher` use PSA (and for that last one, whether GCC and CCM use `block_cipher`).

However, thinking more about it: we don't need to wait until all of the above work is done, or any of it really. The point of the complex definitions for 3.6 is that we need to ensure that the code works without calling `psa_crypto_init()`, except in the circumstances where the documentation says we have to (when `USE_PSA` is on for PK, X.509 and TLS, in certain driver-only builds for MD, CTR-DRBG and GCM/CCM - the details are complicated). That's not something we need in development: we don't need any crypto operation to work without calling `psa_crypto_init()`.

Merging this observation with your proposal to put all the `*PSA_INIT/DONE` behind a single 4.0 guard, we could do immediately  do something like:
```c
#if MBEDTLS_VERSION_MAJOR >= 4
/* Legacy names from 3.6 */
#define USE_PSA_INIT() PSA_INIT()
#define USE_PSA_DONE() PSA_DONE()
#define MD_PSA_INIT()   PSA_INIT()
#define MD_PSA_DONE()   PSA_DONE()
// and so one
#else
// The 3.6 things with all their complexity - but crucially no additional "4.0 vs 3.6" complexity
```

Then at some point in development we can do a general `s/[A-Z_]*PSA_\(INIT\|DONE\)/PSA_\1/` and remove the legacy names - but that would be for another PR.

Wdyt?

**Comment by mpg** [comment]
> Splitting cleanup-related declarations, the decision of calling init, support for exotic options and whatever else would be rather cleaning up a bit.

Indeed, but I think the part of my proposal that adds complexity is when I suggest one of the files continues to live in the main repo for 3.6 and moves to the framework for 4.0 and 1.0 - on second thought, I'm really not happy about that part.

(Without that part, splitting up the file to clean it up becomes quite orthogonal to this PR - would probably still be a good thing, but probably for another PR.)

**Comment by davidhorstmann-arm** [comment]
> Wdyt?

That sounds good to me (and much simpler). I'll give that a try.

**Comment by davidhorstmann-arm** [comment]
ABI-check failing as a result of the recent motion of `error.c` I believe.

**Review by eleuzi01 — APPROVED** [review]
LGTM, thanks!

**Review by mpg — APPROVED** [review]
LGTM, thanks for addressing my comments!
